### PR TITLE
Simulation and QSM tests for `HasBufFS`

### PR DESCRIPTION
**Pull request by jorisdral**
Follow-up to #46

### CODE REVIEW

File: fs-sim/test/Test/System/FS/StateMachine.hs [hunk 1]
```diff
@@ -1467,6 +1593,7 @@ runCmds tmpDir cmds = QC.monadicIO $ do
         $ QSM.checkCommandNames cmds
         $ tabulate "Tags" (map show $ tag (execCmds cmds))
         $ counterexample ("Mount point: " ++ tstTmpDir)
+        $ QSM.checkCommandNames cmds
```
Comment (jasagredo) [inline fs-sim/test/Test/System/FS/StateMachine.hs:1596]: Isn't this one duplicated 3 lines above?

File: fs-sim/test/Test/System/FS/StateMachine.hs [hunk 2]
```diff
@@ -147,7 +156,11 @@ data Cmd fp h =
   | Seek               h SeekMode Int64
   | Get                h Word64
   | GetAt              h Word64 AbsOffset
+  | GetBuf             h ByteCount
+  | GetBufAt           h ByteCount AbsOffset
```
Comment (jasagredo) [inline fs-sim/test/Test/System/FS/StateMachine.hs:160]: One thing I don't like from QSM is that you cannot tell which response is the one expected from each command. Perhaps some comments here might help?
```haskell
  | GetBuf   h ByteCount             -- -> ByteCount
  ...
  | PutBuf   h ByteString ByteCount  -- -> BCBS
```

File: fs-sim/src/System/FS/Sim/MockFS.hs [hunk 3]
```diff
@@ -905,3 +887,235 @@ sign a | a < 0     = Negative (negate a)
...
+          unless success $ throwError (errWritePastBufEnd openFilePath)
+          return (fromIntegral (BS.length bs), hs)
+        Append -> throwError (errNoReadAccess openFilePath "append")
+  where
+    errNoReadAccess fp mode = FsError {
+        fsErrorType   = FsInvalidArgument
+      , fsErrorPath   = fsToFsErrorPathUnmounted fp
+      , fsErrorString = "cannot hGetBufSomeAt in " <> mode <> " mode"
+      , fsErrorNo     = Nothing
+      , fsErrorStack  = prettyCallStack
```
Comment (jasagredo) [inline fs-sim/src/System/FS/Sim/MockFS.hs:1009]: Is it intended that there is no `HasCallStack` constraint on the function? Won't the call stack cut too early?

**Review by jasagredo — APPROVED** [review]

File: fs-sim/test/Test/System/FS/StateMachine.hs (hunk 2 above)
Comment (jorisdral) [reply fs-sim/test/Test/System/FS/StateMachine.hs:160]: Added some comments

File: fs-sim/src/System/FS/Sim/MockFS.hs (hunk 3 above)
Comment (jorisdral) [reply fs-sim/src/System/FS/Sim/MockFS.hs:1009]: The `CanSimFS` tuple includes a `HasCallStack` constraint

File: fs-sim/src/System/FS/Sim/Prim.hs [hunk 4]
```diff
@@ -0,0 +1,98 @@
+{-# LANGUAGE BangPatterns               #-}
+{-# LANGUAGE DataKinds                  #-}
+{-# LANGUAGE DerivingStrategies         #-}
+{-# LANGUAGE GeneralisedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE TypeFamilies               #-}
+{-# LANGUAGE UnboxedTuples              #-}
+{-# LANGUAGE UndecidableInstances       #-}
+
+module System.FS.Sim.Prim (
```
Comment (dnadales) [inline fs-sim/src/System/FS/Sim/Prim.hs:10]: I'm just wondering if this module deserves an explanation :thinking:

**Review by dnadales — APPROVED** [review]
LGTM!

File: fs-sim/src/System/FS/Sim/Prim.hs (hunk 4 above)
Comment (jorisdral) [reply fs-sim/src/System/FS/Sim/Prim.hs:10]: I'll include it in a follow-up!

**Pull request merged**